File: frontend/src/shared/utils/testRunner.ts
// Tests DSL 运行器 — 响应返回后执行测试断言

// 测试 DSL 支持:
//   pm.test("name", () => { ... })
//   pm.response.status         — 状态码
//   pm.response.headers        — 响应头
//   pm.response.body           — 响应体字符串
//   pm.response.json()         — JSON 解析
//   pm.response.time           — 响应时间 (ms)
//   pm.assert.equal(a, b)      — 断言相等
//   pm.assert.ok(value)        — 断言真值
//   pm.assert.contains(a, b)   — 断言包含
//   pm.assert.gt(a, b)         — 断言大于
//   pm.assert.lt(a, b)         — 断言小于

export interface TestCase {
  name: string
  passed: boolean
  message: string
  expected?: any
  actual?: any
  duration?: number
}

export interface TestContext {
  response: {
    status: number
    headers: Record<string, string>
    body: string
    json(): any
    time: number
  }
  request: {
    method: string
    url: string
  }
}

export function runTests(
  script: string,
  context: TestContext,
): TestCase[] {
  const results: TestCase[] = []

  if (!script || !script.trim()) {
    return results
  }

  const pm = {
    test(name: string, fn: () => void) {
      try {
        fn()
        results.push({ name, passed: true, message: '通过' })
      } catch (e: any) {
        results.push({
          name,
          passed: false,
          message: e.message || '测试失败',
        })
      }
    },
    response: context.response,
    request: context.request,
    assert: {
      equal(a: any, b: any, msg?: string) {
        if (a !== b) {
          throw new Error(msg || `期望 ${JSON.stringify(b)}，实际 ${JSON.stringify(a)}`)
        }
      },
      ok(value: any, msg?: string) {
        if (!value) {
          throw new Error(msg || `期望为真值，实际 ${JSON.stringify(value)}`)
        }
      },
      contains(str: string, substr: string, msg?: string) {
        if (!str.includes(substr)) {
          throw new Error(msg || `期望包含 "${substr}"`)
        }
      },
      gt(a: number, b: number, msg?: string) {
        if (!(a > b)) {
          throw new Error(msg || `期望 ${a} > ${b}`)
        }
      },
      lt(a: number, b: number, msg?: string) {
        if (!(a < b)) {
          throw new Error(msg || `期望 ${a} < ${b}`)
        }
      },
    },
  }

  try {
    const fn = new Function('pm', script)
    fn(pm)
  } catch (e: any) {
    results.push({
      name: '脚本执行错误',
      passed: false,
      message: e.message || '未知错误',
    })
  }

  return results
}

File: backend/app/features/request_workbench/test_runner.py
"""Tests 脚本执行引擎 - 后端执行类 JavaScript DSL

在受限沙箱中执行 pm.test/pm.expect/pm.response/pm.environment 语法。
使用 exec() + 受限 globals（__builtins__ 设为空字典）防止沙箱逃逸。
"""


class ExpectChain:
    """链式断言"""

    def __init__(self, value):
        self.value = value

    def toBe(self, expected) -> None:
        """严格相等断言"""
        if self.value != expected:
            raise AssertionError(f"期望 {expected!r}，实际 {self.value!r}")

    def toContain(self, substring) -> None:
        """字符串包含断言"""
        actual_str = str(self.value)
        if substring not in actual_str:
            raise AssertionError(f"期望包含 {substring!r}，实际 {actual_str!r}")

    def toHaveStatus(self, code) -> None:
        """状态码断言"""
        actual = self.value
        if isinstance(actual, dict):
            actual = actual.get("status_code", 0)
        if int(actual) != int(code):
            raise AssertionError(f"期望状态码 {code}，实际 {actual}")

    def toHaveHeader(self, name, value) -> None:
        """响应头断言"""
        headers = self.value
        if isinstance(headers, dict):
            actual = headers.get(name, headers.get(name.lower(), ""))
        else:
            actual = ""
        if str(actual) != str(value):
            raise AssertionError(f"期望响应头 {name}={value!r}，实际 {actual!r}")

    def toBeJson(self) -> None:
        """合法 JSON 断言"""
        import json
        val = self.value
        if isinstance(val, dict):
            try:
                json.dumps(val)
                return
            except Exception:
                pass
        if isinstance(val, str):
            try:
                json.loads(val)
                return
            except json.JSONDecodeError:
                pass
        raise AssertionError(f"不是合法的 JSON: {val!r}")

    def toBeLessThan(self, n) -> None:
        """小于断言"""
        if not (float(self.value) < float(n)):
            raise AssertionError(f"期望小于 {n}，实际 {self.value}")


class PmWrapper:
    """pm.response 对象 — 提供对响应数据的只读访问"""

    def __init__(self, response):
        self._response = response

    @property
    def status(self):
        return self._response.get("status_code", 0)

    @property
    def headers(self):
        return self._response.get("headers", {})

    @property
    def body(self):
        return self._response.get("body", "")

    @property
    def time(self):
        return self._response.get("duration", 0)

    def json(self):
        import json
        body = self._response.get("body", "")
        if isinstance(body, dict):
            return body
        if isinstance(body, str):
            try:
                return json.loads(body)
            except json.JSONDecodeError:
                return body
        return body


class PmEnvironment:
    """pm.environment 对象 — 读写环境变量"""

    def __init__(self, env_vars):
        self._env = env_vars

    def get(self, key):
        return self._env.get(key)

    def set(self, key, value):
        self._env[key] = str(value)


class PmObject:
    """提供给脚本的 pm 对象"""

    def __init__(self, response: dict, env_vars: dict, results_list: list):
        self._response = response
        self._env_vars = env_vars
        self._results = results_list
        self.response = PmWrapper(response)
        self.environment = PmEnvironment(env_vars)

    def test(self, name: str, fn) -> None:
        """执行测试，捕获异常"""
        import time
        start = time.time()
        try:
            fn()
            self._results.append({
                "name": name,
                "passed": True,
                "message": "",
                "expected": None,
                "actual": None,
                "duration": int((time.time() - start) * 1000),
            })
        except AssertionError as e:
            self._results.append({
                "name": name,
                "passed": False,
                "message": str(e),
                "expected": None,
                "actual": None,
                "duration": int((time.time() - start) * 1000),
            })
        except Exception as e:
            self._results.append({
                "name": name,
                "passed": False,
                "message": f"脚本异常: {e}",
                "expected": None,
                "actual": None,
                "duration": int((time.time() - start) * 1000),
            })

    def expect(self, value):
        """返回断言链"""
        return ExpectChain(value)


class TestRunner:
    """在后端执行 Tests 脚本"""

    def __init__(self, response: dict, env_vars: dict = None):
        self.response = response
        self.env_vars = env_vars or {}

    def run(self, script: str) -> list[dict]:
        """执行测试脚本，返回测试结果列表"""
        if not script or not script.strip():
            return []

        results: list[dict] = []
        pm = PmObject(self.response, self.env_vars, results)

        # 构建受限沙箱
        sandbox_globals = {
            "pm": pm,
            "__builtins__": {},
        }

        try:
            exec(script, sandbox_globals)
        except SyntaxError as e:
            results.append({
                "name": "语法错误",
                "passed": False,
                "message": f"脚本语法错误: {e}",
                "expected": None,
                "actual": None,
                "duration": 0,
            })
        except Exception as e:
            results.append({
                "name": "脚本执行错误",
                "passed": False,
                "message": str(e),
                "expected": None,
                "actual": None,
                "duration": 0,
            })

        return results
